<a href="https://colab.research.google.com/github/dscohen75/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [3]:
### YOUR WORK HERE
import pandas as pd


In [10]:
import numpy as np

In [6]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', names=column_headers)

In [7]:
df.shape


(205, 26)

In [8]:
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

## Replacing "?" with NaN:

In [13]:
df = df.replace({"?": np.NaN})

df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [20]:
### YOUR WORK HERE

sample = df.sample(10, random_state = 30)

print(sample['highway-mpg'].mean())  # This is the sample average

32.8


In [21]:
print(df['highway-mpg'].mean()) # This is the presumed population mean

30.75121951219512


One sample t-test comparing the sample mean to the population mean.

HO: The sample mean is equal to the population mean.

H1: The sample mean is NOT equal to the population mean.


In [26]:
from scipy import stats

stats.ttest_1samp(sample['highway-mpg'], 30.75)

Ttest_1sampResult(statistic=0.8917727404948264, pvalue=0.3957307423595724)

## Based upon the t-statistic of 0.89 and p-value of 0.40, we fail to reject the null hypothesis that the sample mean is equal to the population mean.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

## HO: The mean MPG of the sample is equal to 35. 

## H1: The mean MPF of the sample is different from 35.

In [27]:
### YOUR WORK HERE

stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

## Based upon the t-statistic of 0.96 and p-value of 0.36, with this small sample we fail to reject the null hypothesis that the sample mean is equal to the population mean.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [28]:
### YOUR WORK HERE

sample100 = df.sample(100, random_state=30)

stats.ttest_1samp(sample100['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

## With the same null and alternative hypotheses as above, with this larger sample size we now have a t-statistic of -6.5 and p-value of 0.000000003 we can now reject the null hypothesis that the mean of this sample is equal to 35. Furthermore, since the t-statistic is negative, we can say that the true mean MPG of the population from which this sample was drawn is less than 35.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

## The sample size of 10 did not provide enough statistical power to the t-test to detect the difference between the mean of that sample and the hypothesized mean. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [29]:
### YOUR WORK HERE

column_headers2 = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

votes = pd.read_csv('house-votes-84.data', names=column_headers2)

print(votes.shape)
votes.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [30]:
votes = votes.replace({'y':1, 'n':0, '?': np.NaN})

votes.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [39]:
rep = votes[votes['party'] == 'republican']
dem = votes[votes['party'] == 'democrat']

In [34]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [35]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

## We can first look at the means of the votes by issue in each of the dem and rep dataframes, to get an idea of on which issues the two parties seem to differ by a lot:

In [37]:
### YOUR WORK HERE

dem.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [41]:
rep.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

## On the issue of handicapped infants, for example, the means look to be separated by quite a bit.

### H0: The mean reate of republicans voting "yes" for handicapped infants equals the mean reate of democrats voting "yes" for handicapped infants.

### H1: The mean reate of republicans voting "yes" for handicapped infants is different from the mean reate of democrats voting "yes" for handicapped infants.

In [42]:
stats.ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

## With a t-statistic of 9.21 and a p-value of 0.000000000000000002, we easily reject the null hypothesis at the 0.01 level. Since the t-statistic is positive, we can easily say that democratic congresspeople voted yes on handicapped infants at a higher rate than republicans did.

## On the issue of physician fee freeze, the means also look to be separated by quite a bit, in the opposite direction as for handicapped infants. 

### H0: The mean reate of republicans voting "yes" for physician fee freeze equals the mean reate of democrats voting "yes" for handicapped infants.

### H1: The mean reate of republicans voting "yes" for physician fee freeze is different from the mean reate of democrats voting "yes" for handicapped infants.


In [43]:
stats.ttest_ind(dem['physician-fee-freeze'], rep['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

## With a t-statistic of -49.4 and a p-value of 2 x 10^-177, we very easily reject the null hypothesis at the 0.01 level. Since the t-statistic is negative, we can easily say that democratic congresspeople voted yes on handicapped infants at a much lower rate than republicans did.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE

In [65]:
## Define a function that takes a column from the voting dataset
## returns the result of the two-sample t-test comparing
## the mean for democrats to the mean for republicans

def ttest_2samp_column(col):
  return(stats.ttest_ind(dem[str(col)], rep[str(col)], nan_policy='omit'))

Test this function on an issue that we've already run the two-sample t-test on.

In [66]:
ttest_2samp_column('handicapped-infants')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

## The result matches what we got above.

In [68]:
votes.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

## I want to drop the 'party' column first because I know you can't run t-tests on strings.

In [70]:
votes2 = votes.drop(['party'], axis=1)

votes2.head()

handicapped-infants  water-project  budget  ...  crime  duty-free  south-africa
0                  0.0            1.0     0.0  ...    1.0        0.0           1.0
1                  0.0            1.0     0.0  ...    1.0        0.0           NaN
2                  NaN            1.0     1.0  ...    1.0        0.0           0.0
3                  0.0            1.0     1.0  ...    0.0        0.0           1.0
4                  1.0            1.0     1.0  ...    1.0        1.0           1.0

[5 rows x 16 columns]

## I can't figure out why .apply() isn't working for this... please help!

In [71]:
votes2.apply(ttest_2samp_column)

KeyError: ignored



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE